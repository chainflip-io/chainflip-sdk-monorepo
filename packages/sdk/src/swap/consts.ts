export enum ChainId {
  Ethereum = 1,
  Polkadot,
  Bitcoin,
}

export type { SupportedAsset as TokenSymbol } from '@/shared/enums';

// value to be replaced at build time with `envsubst` or similar
export const BACKEND_SERVICE_URL = '$BACKEND_SERVICE_URL';
