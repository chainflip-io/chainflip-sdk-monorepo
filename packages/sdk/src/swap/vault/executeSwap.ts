import assert from 'assert';
import { ContractReceipt, Signer } from 'ethers';
import { z } from 'zod';
import {
  SISYPHOS_FLIP_CONTRACT_ADDRESS,
  GOERLI_USDC_CONTRACT_ADDRESS,
  SISYPHOS_VAULT_CONTRACT_ADDRESS,
} from '@/shared/addresses';
import {
  ChainflipNetwork,
  SupportedAsset,
  chainflipNetwork,
} from '@/shared/enums';
import {
  Vault,
  Vault__factory,
  ERC20__factory,
} from '../../../types/ethers-contracts';
import { ChainId } from '../sdk';
import { isTestnet } from '../utils';
import {
  ExecuteSwapParams,
  NativeSwapParams,
  TokenSwapParams,
  executeSwapParamsSchema,
} from './validators';

// !!!!! IMPORTANT !!!!!
// Do not change these indices.
const chainMap: Record<ChainId, number> = {
  [ChainId.Ethereum]: 1,
  [ChainId.Polkadot]: 2,
  [ChainId.Bitcoin]: 3,
};

// !!!!!! IMPORTANT !!!!!!
// Do not change these indices.
const assetMap: Record<SupportedAsset, number> = {
  // 0 is reservered for particular cross chain messaging scenarios where we want to pass
  // through a message without making a swap.
  ETH: 1,
  FLIP: 2,
  USDC: 3,
  DOT: 4,
  BTC: 5,
};

const swapNative = async (
  vault: Vault,
  { destChainId, destTokenSymbol, destAddress, amount }: NativeSwapParams,
): Promise<ContractReceipt> => {
  const transaction = await vault.xSwapNative(
    chainMap[destChainId],
    destAddress,
    assetMap[destTokenSymbol],
    [],
    { value: amount },
  );

  const receipt = await transaction.wait(1);

  assert(receipt.status !== 0, 'Transaction failed');

  return receipt;
};

const getTokenContractAddress = (
  asset: SupportedAsset,
  cfNetwork: ChainflipNetwork,
): string => {
  assert(isTestnet(cfNetwork), 'Only testnets are supported for now');

  if (asset === 'FLIP' && cfNetwork === 'sisyphos') {
    return SISYPHOS_FLIP_CONTRACT_ADDRESS;
  }

  assert(asset === 'USDC');

  return GOERLI_USDC_CONTRACT_ADDRESS;
};

const swapToken = async (
  vault: Vault,
  params: TokenSwapParams,
  { signer, ...opts }: ExecuteSwapOptions,
): Promise<ContractReceipt> => {
  const erc20Address =
    opts.cfNetwork === 'localnet'
      ? opts.srcTokenContractAddress
      : getTokenContractAddress(params.srcTokenSymbol, opts.cfNetwork);

  assert(erc20Address !== undefined, 'Missing ERC20 contract address');
  const erc20 = ERC20__factory.connect(erc20Address, signer);
  const signerAddress = await signer.getAddress();
  const allowance = await erc20.allowance(signerAddress, vault.address);

  if (allowance.lt(params.amount)) {
    const approval = await erc20.approve(vault.address, params.amount);
    const approvalReceipt = await approval.wait(1);
    assert(approvalReceipt.status !== 0, 'Approval failed');
  }

  const transaction = await vault.xSwapToken(
    chainMap[params.destChainId],
    params.destAddress,
    assetMap[params.destTokenSymbol],
    erc20Address,
    params.amount,
    [],
  );

  const receipt = await transaction.wait(1);

  assert(receipt.status !== 0, 'Transaction failed');

  return receipt;
};

const isTokenSwap = (params: ExecuteSwapParams): params is TokenSwapParams =>
  'srcTokenSymbol' in params;

const executeSwapOptionsSchema = z.intersection(
  z.object({ signer: z.instanceof(Signer) }),
  z.union([
    z.object({ cfNetwork: chainflipNetwork }),
    z.object({
      cfNetwork: z.literal('localnet'),
      vaultContractAddress: z.string(),
      srcTokenContractAddress: z.string().optional(),
    }),
  ]),
);

export type ExecuteSwapOptions = z.infer<typeof executeSwapOptionsSchema>;

const executeSwap = async (
  params: ExecuteSwapParams,
  options: ExecuteSwapOptions,
): Promise<ContractReceipt> => {
  executeSwapParamsSchema.parse(params);
  const opts = executeSwapOptionsSchema.parse(options);

  let vaultContractAddress: string | undefined;
  if (opts.cfNetwork === 'localnet') {
    vaultContractAddress = opts.vaultContractAddress;
  } else if (opts.cfNetwork === 'sisyphos') {
    vaultContractAddress = SISYPHOS_VAULT_CONTRACT_ADDRESS;
  }

  assert(
    vaultContractAddress !== undefined,
    'Missing vault contract address or network unsupported',
  );

  const vault = Vault__factory.connect(vaultContractAddress, opts.signer);

  if (isTokenSwap(params)) return swapToken(vault, params, opts);
  return swapNative(vault, params);
};

export default executeSwap;
