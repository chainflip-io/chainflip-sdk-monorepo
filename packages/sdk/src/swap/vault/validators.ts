import { z } from 'zod';
import { supportedAsset } from '@/shared/enums';
import {
  btcAddress,
  dotAddress,
  ethereumAddress,
  numericString,
} from '@/shared/parsers';
import { ChainId } from '../sdk';

const base = z.object({ amount: numericString });

const ethereumBase = base.extend({
  destChainId: z.literal(ChainId.Ethereum),
  destAddress: ethereumAddress,
});

const polkadotBase = base.extend({
  destChainId: z.literal(ChainId.Polkadot),
  destAddress: dotAddress,
});

const bitcoinBase = base.extend({
  destChainId: z.literal(ChainId.Bitcoin),
  destAddress: btcAddress,
});

const erc20 = z.union([
  z.literal(supportedAsset.enum.FLIP),
  z.literal(supportedAsset.enum.USDC),
]);

const ethereumNative = ethereumBase
  .extend({
    destTokenSymbol: erc20,
  })
  .strict();

const polkadotNative = polkadotBase
  .extend({
    destTokenSymbol: z.literal(supportedAsset.enum.DOT),
    srcTokenSymbol: z.undefined(),
  })
  .strict();

const bitcoinNative = bitcoinBase
  .extend({
    destTokenSymbol: z.literal(supportedAsset.enum.BTC),
    srcTokenSymbol: z.undefined(),
  })
  .strict();

const nativeSwapParamsSchema = z.union([
  ethereumNative,
  polkadotNative,
  bitcoinNative,
]);

export type NativeSwapParams = z.infer<typeof nativeSwapParamsSchema>;

const tokenSwapParamsSchema = z.union([
  ethereumBase.extend({
    srcTokenSymbol: z.literal(supportedAsset.enum.FLIP),
    destTokenSymbol: z.union([
      z.literal(supportedAsset.enum.USDC),
      z.literal(supportedAsset.enum.ETH),
    ]),
  }),
  ethereumBase.extend({
    srcTokenSymbol: z.literal(supportedAsset.enum.USDC),
    destTokenSymbol: z.union([
      z.literal(supportedAsset.enum.FLIP),
      z.literal(supportedAsset.enum.ETH),
    ]),
  }),
  polkadotNative.extend({ srcTokenSymbol: erc20 }),
  bitcoinNative.extend({ srcTokenSymbol: erc20 }),
]);

export type TokenSwapParams = z.infer<typeof tokenSwapParamsSchema>;

export const executeSwapParamsSchema = z.union([
  nativeSwapParamsSchema,
  tokenSwapParamsSchema,
]);

export type ExecuteSwapParams = z.infer<typeof executeSwapParamsSchema>;
