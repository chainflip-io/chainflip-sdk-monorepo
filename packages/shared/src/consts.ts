import { Asset, Assets, ChainflipNetwork, ChainflipNetworks } from './enums';

const POOLS_NETWORK_FEE_HUNDREDTH_PIPS: Partial<
  Record<ChainflipNetwork, number>
> = {
  [ChainflipNetworks.perseverance]: 100,
};
export const getPoolsNetworkFeeHundredthPips = (network: ChainflipNetwork) =>
  POOLS_NETWORK_FEE_HUNDREDTH_PIPS[network] ?? 0;

// TODO: fetch minimum deposit amounts via rpc from the state chain
const MINIMUM_DEPOSIT_AMOUNTS: Partial<
  Record<ChainflipNetwork, Record<Asset, string>>
> = {
  [ChainflipNetworks.perseverance]: {
    [Assets.ETH]: '0',
    [Assets.FLIP]: '0',
    [Assets.USDC]: '0',
    [Assets.BTC]: '0',
    [Assets.DOT]: '10000000000', // https://support.polkadot.network/support/solutions/articles/65000168651-what-is-the-existential-deposit
  },
};
export const getMinimumDepositAmount = (
  network: ChainflipNetwork,
  asset: Asset,
) => MINIMUM_DEPOSIT_AMOUNTS[network]?.[asset] ?? '0';

// TODO: fetch minimum swap amounts via rpc from the state chain
const MINIMUM_SWAP_AMOUNTS: Partial<
  Record<ChainflipNetwork, Record<Asset, string>>
> = {
  [ChainflipNetworks.perseverance]: {
    [Assets.ETH]: '580000000000000',
    [Assets.FLIP]: '1000000000000000000',
    [Assets.USDC]: '1000000',
    [Assets.BTC]: '1000',
    [Assets.DOT]: '2000000000',
  },
};
export const getMinimumSwapAmount = (network: ChainflipNetwork, asset: Asset) =>
  MINIMUM_SWAP_AMOUNTS[network]?.[asset] ?? '0';

export const ADDRESSES = {
  [ChainflipNetworks.sisyphos]: {
    FLIP_CONTRACT_ADDRESS: '0x2BbB561C6eaB74f358cA9e8a961E3A20CAE3D100',
    VAULT_CONTRACT_ADDRESS: '0xC17CCec5015081EB2DF26d20A9e02c5484C1d641',
    STATE_CHAIN_GATEWAY_ADDRESS: '0xE8bE4B7F8a38C1913387c9C20B94402bc3Db9F70',
  },
  [ChainflipNetworks.perseverance]: {
    FLIP_CONTRACT_ADDRESS: '0x0485D65da68b2A6b48C3fA28D7CCAce196798B94',
    VAULT_CONTRACT_ADDRESS: '0x40caFF3f3B6706Da904a7895e0fC7F7922437e9B',
    STATE_CHAIN_GATEWAY_ADDRESS: '0x38AA40B7b5a70d738baBf6699a45DacdDBBEB3fc',
  },
} as const;

// https://developers.circle.com/developer/docs/usdc-on-testnet#usdc-on-ethereum-goerli
export const GOERLI_USDC_CONTRACT_ADDRESS =
  '0x07865c6E87B9F70255377e024ace6630C1Eaa37F';
