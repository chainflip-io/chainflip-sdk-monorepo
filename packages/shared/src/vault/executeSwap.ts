import { ContractReceipt } from 'ethers';
import { Vault__factory } from '../abis';
import {
  checkAllowance,
  getTokenContractAddress,
  getVaultManagerContractAddress,
} from '../contracts';
import { assetContractIds, chainContractIds } from '../enums';
import { assert, isTokenCall, isTokenSwap } from '../guards';
import {
  executeOptionsSchema,
  type ExecuteOptions,
  type ExecuteSwapParams,
  executeSwapParamsSchema,
  type NativeSwapParams,
  type TokenSwapParams,
  NativeCallParams,
  TokenCallParams,
} from './schemas';

const swapNative = async (
  { destChain, destAsset, destAddress, amount }: NativeSwapParams,
  { network, vaultContractAddress: address, signer, ...opts }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet' ? address : getVaultManagerContractAddress(network);

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xSwapNative(
    chainContractIds[destChain],
    destAddress,
    assetContractIds[destAsset],
    [],
    { value: amount, ...opts },
  );

  return transaction.wait(1);
};

const swapToken = async (
  params: TokenSwapParams,
  {
    network,
    vaultContractAddress: vaultAddress,
    srcTokenContractAddress: tokenAddress,
    signer,
    ...opts
  }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet'
      ? vaultAddress
      : getVaultManagerContractAddress(network);

  const erc20Address =
    network === 'localnet'
      ? tokenAddress
      : getTokenContractAddress(params.srcAsset, network);

  assert(erc20Address !== undefined, 'Missing ERC20 contract address');

  const { isAllowable } = await checkAllowance(
    params.amount,
    vaultContractAddress,
    erc20Address,
    signer,
  );
  assert(isAllowable, 'Swap amount exceeds allowance');

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xSwapToken(
    chainContractIds[params.destChain],
    params.destAddress,
    assetContractIds[params.destAsset],
    erc20Address,
    params.amount,
    [],
    opts,
  );

  return transaction.wait(1);
};

const callNative = async (
  params: NativeCallParams,
  {
    network,
    vaultContractAddress: vaultAddress,
    signer,
    ...opts
  }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet'
      ? vaultAddress
      : getVaultManagerContractAddress(network);

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xCallNative(
    chainContractIds[params.destChain],
    params.destAddress,
    assetContractIds[params.destAsset],
    params.ccmMetadata.message,
    params.ccmMetadata.gasBudget,
    [],
    { value: params.amount, ...opts },
  );

  return transaction.wait(1);
};

const callToken = async (
  params: TokenCallParams,
  {
    signer,
    network,
    vaultContractAddress: vaultAddress,
    srcTokenContractAddress: tokenAddress,
    ...opts
  }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet'
      ? vaultAddress
      : getVaultManagerContractAddress(network);

  const erc20Address =
    network === 'localnet'
      ? tokenAddress
      : getTokenContractAddress(params.srcAsset, network);

  assert(erc20Address !== undefined, 'Missing ERC20 contract address');

  const { isAllowable } = await checkAllowance(
    params.amount,
    vaultContractAddress,
    erc20Address,
    signer,
  );
  assert(isAllowable, 'Swap amount exceeds allowance');

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xCallToken(
    chainContractIds[params.destChain],
    params.destAddress,
    assetContractIds[params.destAsset],
    params.ccmMetadata.message,
    params.ccmMetadata.gasBudget,
    erc20Address,
    params.amount,
    [],
    opts,
  );

  return transaction.wait(1);
};

const executeSwap = async (
  params: ExecuteSwapParams,
  options: ExecuteOptions,
): Promise<ContractReceipt> => {
  const parsedParams = executeSwapParamsSchema.parse(params);
  const opts = executeOptionsSchema.parse(options);

  if ('ccmMetadata' in parsedParams) {
    return isTokenCall(parsedParams)
      ? callToken(parsedParams, opts)
      : callNative(parsedParams, opts);
  }

  return isTokenSwap(parsedParams)
    ? swapToken(parsedParams, opts)
    : swapNative(parsedParams, opts);
};

export default executeSwap;
