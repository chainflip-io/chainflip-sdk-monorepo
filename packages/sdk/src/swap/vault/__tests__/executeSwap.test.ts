/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/lines-between-class-members */
/* eslint-disable max-classes-per-file */
import { BigNumber, VoidSigner } from 'ethers';
import { ChainId } from '../../consts';
import executeSwap from '../executeSwap';
import { ExecuteSwapParams } from '../validators';

const ETH_ADDRESS = '0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2';
const DOT_ADDRESS = '5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX';
const BTC_ADDRESS = 'bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq';

class MockVault {
  constructor(readonly address: string) {}
  async xSwapNative(): Promise<any> {}
  async xSwapToken(): Promise<any> {}
}

class MockERC20 {
  async approve(): Promise<any> {}
  async allowance(): Promise<any> {
    return BigNumber.from(Number.MAX_SAFE_INTEGER - 1);
  }
}

jest.mock(
  '../../../../types/ethers-contracts/factories/Vault__factory',
  () => ({
    Vault__factory: class {
      static connect: (address: string) => MockVault = jest.fn(
        (address: string) => new MockVault(address),
      );
    },
  }),
);

jest.mock(
  '../../../../types/ethers-contracts/factories/ERC20__factory',
  () => ({
    ERC20__factory: class {
      static connect: () => MockERC20 = jest.fn(() => new MockERC20());
    },
  }),
);

describe(executeSwap, () => {
  it.each(['perseverance', 'mainnet'] as const)(
    'only works on sisyphos for now',
    async (cfNetwork) => {
      await expect(
        executeSwap({} as any, {
          cfNetwork,
          signer: new VoidSigner('MY ADDRESS'),
        }),
      ).rejects.toThrowError();
    },
  );

  it.each([
    {
      destTokenSymbol: 'BTC',
      destChainId: ChainId.Bitcoin,
      destAddress: BTC_ADDRESS,
    },
    {
      destTokenSymbol: 'FLIP',
      destChainId: ChainId.Ethereum,
      destAddress: ETH_ADDRESS,
    },
    {
      destTokenSymbol: 'USDC',
      destChainId: ChainId.Ethereum,
      destAddress: ETH_ADDRESS,
    },
    {
      destTokenSymbol: 'DOT',
      destChainId: ChainId.Polkadot,
      destAddress: DOT_ADDRESS,
    },
  ] as Omit<ExecuteSwapParams, 'amount'>[])(
    'submits a native swap (%p)',
    async (params) => {
      const wait = jest
        .fn()
        .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
      const swapSpy = jest
        .spyOn(MockVault.prototype, 'xSwapNative')
        .mockResolvedValue({ wait });

      expect(
        await executeSwap({ amount: '1', ...params } as ExecuteSwapParams, {
          cfNetwork: 'sisyphos',
          signer: new VoidSigner('MY ADDRESS'),
        }),
      ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
      expect(wait).toHaveBeenCalledWith(1);
      expect(swapSpy.mock.calls).toMatchSnapshot();
    },
  );

  it('throws an error if the transaction errors', async () => {
    const wait = jest.fn().mockResolvedValue({ status: 0 });
    jest.spyOn(MockVault.prototype, 'xSwapNative').mockResolvedValue({ wait });

    await expect(
      executeSwap(
        {
          destTokenSymbol: 'DOT',
          destChainId: ChainId.Polkadot,
          destAddress: DOT_ADDRESS,
          amount: '1',
        },
        { cfNetwork: 'sisyphos', signer: new VoidSigner('MY ADDRESS') },
      ),
    ).rejects.toThrowError();
  });

  it.each([
    ...['FLIP', 'USDC'].flatMap((srcTokenSymbol) => [
      {
        destTokenSymbol: 'BTC',
        destChainId: ChainId.Bitcoin,
        destAddress: BTC_ADDRESS,
        srcTokenSymbol,
      },
      {
        destTokenSymbol: 'ETH',
        destChainId: ChainId.Ethereum,
        destAddress: ETH_ADDRESS,
        srcTokenSymbol,
      },
      {
        destTokenSymbol: 'DOT',
        destChainId: ChainId.Polkadot,
        destAddress: DOT_ADDRESS,
        srcTokenSymbol,
      },
    ]),
  ] as Omit<ExecuteSwapParams, 'amount'>[])(
    'submits a token swap (%p)',
    async (params) => {
      const wait = jest
        .fn()
        .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
      const approveSpy = jest
        .spyOn(MockERC20.prototype, 'approve')
        .mockResolvedValue({ wait });
      const swapSpy = jest
        .spyOn(MockVault.prototype, 'xSwapToken')
        .mockResolvedValue({ wait });
      const allowanceSpy = jest.spyOn(MockERC20.prototype, 'allowance');

      expect(
        await executeSwap({ amount: '1', ...params } as ExecuteSwapParams, {
          cfNetwork: 'sisyphos',
          signer: new VoidSigner('MY ADDRESS'),
        }),
      ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
      expect(wait).toHaveBeenCalledWith(1);
      expect(swapSpy.mock.calls).toMatchSnapshot();
      expect(allowanceSpy.mock.calls).toMatchSnapshot();
      expect(approveSpy).not.toHaveBeenCalled();
    },
  );

  it('submits a token swap with approval', async () => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
    const approveSpy = jest
      .spyOn(MockERC20.prototype, 'approve')
      .mockResolvedValue({ wait });
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapToken')
      .mockResolvedValue({ wait });
    const allowanceSpy = jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigNumber.from(0));

    expect(
      await executeSwap(
        {
          destTokenSymbol: 'BTC',
          destChainId: ChainId.Bitcoin,
          destAddress: BTC_ADDRESS,
          srcTokenSymbol: 'FLIP',
          amount: '1',
        } as ExecuteSwapParams,
        { cfNetwork: 'sisyphos', signer: new VoidSigner('MY ADDRESS') },
      ),
    ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(1);
    expect(swapSpy.mock.calls).toMatchSnapshot();
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
    expect(approveSpy.mock.calls).toMatchSnapshot();
  });

  it('throws if the approve transaction errors', async () => {
    const wait = jest.fn().mockResolvedValue({ status: 0 });
    jest.spyOn(MockERC20.prototype, 'approve').mockResolvedValue({ wait });
    jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigNumber.from(0));

    await expect(
      executeSwap(
        {
          destTokenSymbol: 'BTC',
          destChainId: ChainId.Bitcoin,
          destAddress: BTC_ADDRESS,
          srcTokenSymbol: 'FLIP',
          amount: '1',
        },
        { cfNetwork: 'sisyphos', signer: new VoidSigner('MY ADDRESS') },
      ),
    ).rejects.toThrowError();
  });

  it('throws if the swap transaction errors', async () => {
    const wait = jest.fn().mockResolvedValueOnce({ status: 0 });
    jest.spyOn(MockERC20.prototype, 'approve').mockResolvedValue({ wait });
    jest.spyOn(MockVault.prototype, 'xSwapToken').mockResolvedValue({ wait });

    await expect(
      executeSwap(
        {
          destTokenSymbol: 'BTC',
          destChainId: ChainId.Bitcoin,
          destAddress: BTC_ADDRESS,
          srcTokenSymbol: 'FLIP',
          amount: '1',
        },
        { cfNetwork: 'sisyphos', signer: new VoidSigner('MY ADDRESS') },
      ),
    ).rejects.toThrowError();
  });

  it('can be invoked with localnet options', async () => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
    const approveSpy = jest
      .spyOn(MockERC20.prototype, 'approve')
      .mockResolvedValue({ wait });
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapToken')
      .mockResolvedValue({ wait });
    const allowanceSpy = jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigNumber.from(0));

    expect(
      await executeSwap(
        {
          destTokenSymbol: 'BTC',
          destChainId: ChainId.Bitcoin,
          destAddress: BTC_ADDRESS,
          srcTokenSymbol: 'FLIP',
          amount: '1',
        } as ExecuteSwapParams,
        {
          cfNetwork: 'localnet',
          signer: new VoidSigner('MY ADDRESS'),
          vaultContractAddress: '0x123',
          srcTokenContractAddress: '0x456',
        },
      ),
    ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(1);
    expect(swapSpy.mock.calls).toMatchSnapshot();
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
    expect(approveSpy.mock.calls).toMatchSnapshot();
  });
});
