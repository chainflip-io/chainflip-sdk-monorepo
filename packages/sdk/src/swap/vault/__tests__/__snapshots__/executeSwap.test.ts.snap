// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executeSwap can be invoked with localnet options 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x456",
    "1",
    [],
  ],
]
`;

exports[`executeSwap can be invoked with localnet options 2`] = `
[
  [
    "MY ADDRESS",
    "0x123",
  ],
]
`;

exports[`executeSwap can be invoked with localnet options 3`] = `
[
  [
    "0x123",
    "1",
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destChainId": 1, "destTokenSymbol": "FLIP"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    2,
    [],
    {
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destChainId": 1, "destTokenSymbol": "USDC"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    3,
    [],
    {
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destChainId": 2, "destTokenSymbol": "DOT"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    [],
    {
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destChainId": 3, "destTokenSymbol": "BTC"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destChainId": 1, "destTokenSymbol": "ETH", "srcTokenSymbol": "FLIP"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    1,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destChainId": 1, "destTokenSymbol": "ETH", "srcTokenSymbol": "FLIP"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destChainId": 1, "destTokenSymbol": "ETH", "srcTokenSymbol": "USDC"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    1,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destChainId": 1, "destTokenSymbol": "ETH", "srcTokenSymbol": "USDC"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destChainId": 2, "destTokenSymbol": "DOT", "srcTokenSymbol": "FLIP"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destChainId": 2, "destTokenSymbol": "DOT", "srcTokenSymbol": "FLIP"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destChainId": 2, "destTokenSymbol": "DOT", "srcTokenSymbol": "USDC"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destChainId": 2, "destTokenSymbol": "DOT", "srcTokenSymbol": "USDC"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destChainId": 3, "destTokenSymbol": "BTC", "srcTokenSymbol": "FLIP"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destChainId": 3, "destTokenSymbol": "BTC", "srcTokenSymbol": "FLIP"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destChainId": 3, "destTokenSymbol": "BTC", "srcTokenSymbol": "USDC"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destChainId": 3, "destTokenSymbol": "BTC", "srcTokenSymbol": "USDC"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap with approval 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
  ],
]
`;

exports[`executeSwap submits a token swap with approval 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap with approval 3`] = `
[
  [
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
    "1",
  ],
]
`;
